### PR TITLE
build: add nix flake-based development environment

### PR DESCRIPTION
**Pull request by shanesveller**
Please see the included README.md for additional context. Adopting any of this tooling for building a CI docker image, or using Nix directly in CI, is out of scope for this particular PR as it will require a non-trivial caching strategy for `/nix/store` data. We are starting with the local story first.

## Current behavior

We currently supply configurations for GitHub Actions and a VSCode [devcontainer](https://github.com/build-trust/ockam/blob/develop/.devcontainer/Dockerfile) that is based on our [builder Docker image](https://github.com/build-trust/ockam/blob/8df7833324e2a16a81f2d75bd55960ad7064fcc4/tools/docker/builder/Dockerfile). The project does not currently include any other programmatic definition of a development environment for any of our implementations.

## Proposed changes

This PR introduces an experimental/unsupported Nix flake that allows declarative, versioned, reproducible, cross-platform development environments from a shared source of truth. In the future it could conceivably replace the Dockerfile approach for our builder image, or even replace the Docker-based context. As above, please see the PR's included README file for more context.

Implementation References:
- https://github.com/hercules-ci/flake-parts
- https://flake.parts/define-custom-flake-attribute.html
- https://flake.parts/best-practices-for-module-writing.html
- https://github.com/shanesveller/nix-beam-flakes
- https://github.com/oxalica/rust-overlay
- https://nixos.org/manual/nix/stable/language/builtins.html
- https://ryantm.github.io/nixpkgs/functions/library/options/#sec-functions-library-options
- https://nixos.wiki/wiki/NixOS_modules
- https://nixos.org/manual/nixpkgs/stable/#chap-overlays

New-to-Nix resources:
- https://zero-to-nix.com/
- https://github.com/DeterminateSystems/nix-installer

## Checks

- [X] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [X] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [X] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [X] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

### CODE REVIEW

**Comment by mrinalwadhwa** [comment]
With a couple of minor changes that I've pushed to this branch
https://github.com/build-trust/ockam/tree/mrinal/shanesveller/nix-flake

The following succeed for me, rebased to the latest code in develop, in nix

```
./gradlew build
./gradlew test

CI=true ./gradlew test
```

**Comment by shanesveller** [comment]
@mrinalwadhwa I had mistakenly created this branch off of `main` instead of `develop`, which is how I arrived at some of the archaic versions in the original commits. I've rebased and cherry-picked your logic fix for `DYLD_FALLBACK_LIBRARY_PATH`.

File: tools/nix/README.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# Nix
+
+This prototype [Nix Flake](https://zero-to-nix.com/concepts/flakes) offers a
+self-contained, versioned, reproducible development environment for working on
+the Elixir/Rust libraries offered within this open-source repository. This flake
+is highly experimental with no stability or suitability guarantees, and not
+encouraged for use by external contributors.
```
Comment (mrinalwadhwa) [inline tools/nix/README.md]: > This flake is highly experimental with no stability or suitability guarantees, and not encouraged for use by external contributors.

We can remove this.

File: tools/nix/README.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+## Aside: Nix Use Within Ockam
+
+A few individuals within the Ockam engineering team use Nix in some personal
+capacity. Ockam is exploring use of Nix in a more intentional capacity but at
+this time we have no formal plans or timeline to share regarding this subject.
```
Comment (mrinalwadhwa) [inline tools/nix/README.md]: We can remove this part.

**Review by mrinalwadhwa — DISMISSED** [review]

**Comment by mrinalwadhwa** [comment]
@shanesveller I think we can land this. Note I left a couple of minor comments to remove some caveats. With that change, let's squash the commits and land.

Can't wait to start using it ... there are a couple of open PRs from @glenngillen where new tools are being added and I'd love to start using/improving it there.

File: tools/nix/README.md (hunk 1 above)
Comment (BeenzSyed) [reply tools/nix/README.md]: I recommend to remove this later when we have user interest and can prioritize efforts around maintainability of it.

**Review by mrinalwadhwa — APPROVED** [review]

**Pull request merged**